### PR TITLE
[fix](join) core caused by null-safe-equal join

### PR DESCRIPTION
**Pull request by mrhhsg**
## Proposed changes
```
*** Check failure stack trace: ***
F20240321 11:59:37.800169 12514 column.h:607] Column Nullable(Int64) is not a contiguous block of memory
```
```
 0# doris::signal::(anonymous namespace)::FailureSignalHandler(int, siginfo_t*, void*) at /root/doris/be/src/common/signal_handler.h:421
 1# 0x00007F6B6DC38090 in /lib/x86_64-linux-gnu/libc.so.6
 2# raise at ../sysdeps/unix/sysv/linux/raise.c:51
 3# abort at /build/glibc-SzIz7B/glibc-2.31/stdlib/abort.c:81
 4# 0x000055623D6DBC5D in /root/deploy/be/lib/doris_be
 5# 0x000055623D6CE2FA in /root/deploy/be/lib/doris_be
 6# google::LogMessage::SendToLog() in /root/deploy/be/lib/doris_be
 7# google::LogMessage::Flush() in /root/deploy/be/lib/doris_be
 8# google::LogMessageFatal::~LogMessageFatal() in /root/deploy/be/lib/doris_be
 9# doris::vectorized::IColumn::get_raw_data() const in /root/deploy/be/lib/doris_be
10# void doris::vectorized::MethodKeysFixed >, false>::pack_fixeds(unsigned long, std::vector > const&, std::vector > const&, std::vector >&) at /root/doris/be/src/vec/common/hash_table/hash_map_context.h:404
11# doris::vectorized::MethodKeysFixed >, false>::init_serialized_keys(std::vector > const&, unsigned long, unsigned char const*, bool, bool, unsigned int) at /root/doris/be/src/vec/common/hash_table/hash_map_context.h:469
12# doris::Status doris::vectorized::ProcessHashTableBuild >, false>, doris::vectorized::HashJoinNode>::run<1, false, false, true>(doris::vectorized::MethodKeysFixed >, false>&, doris::vectorized::PODArray, 15ul, 16ul> const*, bool*) at /root/doris/be/src/vec/exec/join/vhash_join_node.h:129
13# doris::Status doris::vectorized::HashJoinNode::_process_build_block(doris::RuntimeState*, doris::vectorized::Block&)::$_1::operator() >, false>&, std::integral_constant&, std::integral_constant, std::integral_constant, std::integral_constant >(doris::vectorized::MethodKeysFixed >, false>&, std::integral_constant&, std::integral_constant, std::integral_constant, std::integral_constant) const at /root/doris/be/src/vec/exec/join/vhash_join_node.cpp:988
14# doris::Status std::__invoke_impl, doris::vectorized::MethodKeysFixed >, false>&, std::integral_constant&, std::integral_constant, std::integral_constant, std::integral_constant >(std::__invoke_other, doris::vectorized::Overload&&, doris::vectorized::MethodKeysFixed >, false>&, std::integral_constant&, std::integral_constant&&, std::integral_constant&&, std::integral_constant&&) at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:61
15# std::__invoke_result, doris::vectorized::MethodKeysFixed >, false>&, std::integral_constant&, std::integral_constant, std::integral_constant, std::integral_constant >::type std::__invoke, doris::vectorized::MethodKeysFixed >, false>&, std::integral_constant&, std::integral_constant, std::integral_constant, std::integral_constant >(doris::vectorized::Overload&&, doris::vectorized::MethodKeysFixed >, false>&, std::integral_constant&, std::integral_constant&&, std::integral_constant&&, std::integral_constant&&) at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:96
16# _ZNSt8__detail9__variant17__gen_vtable_implINS0_12_Multi_arrayIPFNS0_21__deduce_visit_resultIN5doris6StatusEEEONS4_10vectorized8OverloadIJZNS7_12HashJoinNode20_process_build_blockEPNS4_12RuntimeStateERNS7_5BlockEE3$_0ZNS9_20_process_build_blockESB_SD_E3$_1EEERSt7variantIJSt9monostateNS7_16MethodSerializedINS4_13JoinHashTableINS4_9StringRefE11DefaultHashISM_vEEEEENS7_15MethodOneNumberIhNSL_Ih9HashCRC32IhEEEEENSR_ItNSL_ItSS_ItEEEEENSR_IjNSL_IjSS_IjEEEEENSR_ImNSL_ImSS_ImEEEEENSR_INS7_7UInt128ENSL_IS15_SS_IS15_EEEEENSR_INS7_7UInt256ENSL_IS19_SS_IS19_EEEEENS7_15MethodKeysFixedIS13_Lb1EEENS1D_IS13_Lb0EEENS1D_IS17_Lb1EEENS1D_IS17_Lb0EEENS1D_IS1B_Lb1EEENS1D_IS1B_Lb0EEENS1D_INSL_INS7_7UInt136ESS_IS1K_EEELb1EEENS1D_IS1M_Lb0EEEEERSI_IJSt17integral_constantINS4_7TJoinOp4typeELS1T_0EES1R_IS1T_LS1T_2EES1R_IS1T_LS1T_8EES1R_IS1T_LS1T_1EES1R_IS1T_LS1T_4EES1R_IS1T_LS1T_3EES1R_IS1T_LS1T_5EES1R_IS1T_LS1T_7EES1R_IS1T_LS1T_9EES1R_IS1T_LS1T_10EES1R_IS1T_LS1T_11EEEEOSI_IJS1R_IbLb0EES1R_IbLb1EEEES2A_S2A_EJEEESt16integer_sequenceImJLm9ELm3ELm0ELm0ELm1EEEE14__visit_invokeESH_S1Q_S26_S2A_S2A_S2A_ at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/variant:1013
17# decltype(auto) std::__do_visit, doris::vectorized::Overload, std::variant > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false>, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false>, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false>, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false> >&, std::variant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant >&, std::variant, std::integral_constant >, std::variant, std::integral_constant >, std::variant, std::integral_constant > >(doris::vectorized::Overload&&, std::variant > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false>, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false>, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false>, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false> >&, std::variant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant >&, std::variant, std::integral_constant >&&, std::variant, std::integral_constant >&&, std::variant, std::integral_constant >&&) at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/variant:1714
18# decltype(auto) std::visit, std::variant > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false>, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false>, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false>, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false> >&, std::variant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant >&, std::variant, std::integral_constant >, std::variant, std::integral_constant >, std::variant, std::integral_constant > >(doris::vectorized::Overload&&, std::variant > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodOneNumber > >, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false>, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false>, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false>, doris::vectorized::MethodKeysFixed >, true>, doris::vectorized::MethodKeysFixed >, false> >&, std::variant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant, std::integral_constant >&, std::variant, std::integral_constant >&&, std::variant, std::integral_constant >&&, std::variant, std::integral_constant >&&) at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/variant:1769
19# doris::vectorized::HashJoinNode::_process_build_block(doris::RuntimeState*, doris::vectorized::Block&) at /root/doris/be/src/vec/exec/join/vhash_join_node.cpp:971
20# doris::vectorized::HashJoinNode::sink(doris::RuntimeState*, doris::vectorized::Block*, bool) at /root/doris/be/src/vec/exec/join/vhash_join_node.cpp:759
21# doris::pipeline::StreamingOperator::sink(doris::RuntimeState*, doris::vectorized::Block*, doris::pipeline::SourceState) at /root/doris/be/src/pipeline/exec/operator.h:334
22# doris::pipeline::PipelineTask::execute(bool*) at /root/doris/be/src/pipeline/pipeline_task.cpp:304
23# doris::pipeline::TaskScheduler::_do_work(unsigned long) at /root/doris/be/src/pipeline/task_scheduler.cpp:339
24# doris::pipeline::TaskScheduler::start()::$_0::operator()() const at /root/doris/be/src/pipeline/task_scheduler.cpp:218
25# void std::__invoke_impl(std::__invoke_other, doris::pipeline::TaskScheduler::start()::$_0&) at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:61
26# std::enable_if, void>::type std::__invoke_r(doris::pipeline::TaskScheduler::start()::$_0&) at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:117
27# std::_Function_handler::_M_invoke(std::_Any_data const&) at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:291
28# std::function::operator()() const at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:560
29# doris::FunctionRunnable::run() at /root/doris/be/src/util/threadpool.cpp:48
30# doris::ThreadPool::dispatch_thread() at /root/doris/be/src/util/threadpool.cpp:543
31# void std::__invoke_impl(std::__invoke_memfun_deref, void (doris::ThreadPool::*&)(), doris::ThreadPool*&) at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:74
32# std::__invoke_result::type std::__invoke(void (doris::ThreadPool::*&)(), doris::ThreadPool*&) at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:96
33# void std::_Bind::__call(std::tuple<>&&, std::_Index_tuple<0ul>) at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/functional:420
34# void std::_Bind::operator()<, void>() at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/functional:503
35# void std::__invoke_impl&>(std::__invoke_other, std::_Bind&) at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:61
36# std::enable_if&>, void>::type std::__invoke_r&>(std::_Bind&) at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/invoke.h:117
37# std::_Function_handler >::_M_invoke(std::_Any_data const&) at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:291
38# std::function::operator()() const at /var/local/ldb_toolchain/bin/../lib/gcc/x86_64-linux-gnu/11/../../../../include/c++/11/bits/std_function.h:560
39# doris::Thread::supervise_thread(void*) at /root/doris/be/src/util/thread.cpp:498
40# start_thread at /build/glibc-SzIz7B/glibc-2.31/nptl/pthread_create.c:478
41# __clone at ../sysdeps/unix/sysv/linux/x86_64/clone.S:97
```

## Further comments

If this is a relatively large or complex change, kick off the discussion at [dev@doris.apache.org](mailto:dev@doris.apache.org) by explaining why you chose the solution you did and what alternatives you considered, etc...

### CODE REVIEW

**Comment by mrhhsg** [comment]
run buildall

File: be/src/vec/common/hash_table/hash_map_context_creator.h [hunk 1]
```diff
@@ -85,4 +85,36 @@ bool try_get_hash_map_context_fixed(Variant& variant, const VExprContextSPtrs& e
     return use_fixed_key;
 }
 
+template <template <typename... Args> typename HashMap, template <typename> typename Hash,
+          typename... Mapped, typename Variant>
+bool try_get_hash_map_context_fixed(Variant& variant, const std::vector<DataTypePtr>& data_types) {
```
Comment (HappenLee) [inline be/src/vec/common/hash_table/hash_map_context_creator.h:90]: should del the old api

File: be/src/pipeline/exec/hashjoin_build_sink.cpp [hunk 2]
```diff
@@ -361,8 +361,19 @@ void HashJoinBuildSinkLocalState::_hash_table_init(RuntimeState* state) {
                     }
                     return;
                 }
+
+                std::vector<vectorized::DataTypePtr> data_types;
```
Comment (HappenLee) [inline be/src/pipeline/exec/hashjoin_build_sink.cpp:365]: add comment for the change

**Comment by mrhhsg** [comment]
run buildall

**Review by HappenLee — APPROVED** [review]
LGTM

**Review by yiguolei — APPROVED** [review]

**Pull request merged**